modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,101 +5,167 @@
+      if (hasChanged(data.similarFalseState, match.isSome())) {
+        const node = match.getOr(elm);
+        Arr.each(data.similarFalseCallbacks, (callback) => callback(match.isSome(), { node, format, parents }));
       }
+    }
 
-      if (MatchFormat.matchesUnInheritedFormatSelector(editor, node, format)) {
-        return false;
+    Arr.each(data.others, (item) => {
+      const match = matchingNode(editor, parents, format, item.similar, item.vars);
+      if (hasChanged(item.state, match.isSome())) {
+        const node = match.getOr(elm);
+        item.callback(match.isSome(), { node, format, parents });
       }
     });
   });
-
-  // Check if current formats still match
-  const remainingFormats = filterRemainingFormats(currentFormats.get(), matchedFormats, elm, parents);
-
-  // Update the current formats
-  currentFormats.set({
-    ...newFormats,
-    ...remainingFormats
-  });
 };
 
-const filterRemainingFormats = (currentFormats: FormatCallbacks, matchedFormats: FormatCallbacks, elm: Element, parents: Node[]) => Obj.bifilter(currentFormats, (callbacks: FormatChangeCallback[], format: string) => {
-  if (!Obj.has(matchedFormats, format)) {
-    // Execute callbacks
-    Arr.each(callbacks, (callback: FormatChangeCallback) => {
-      callback(false, { node: elm, format, parents });
-    });
-
-    return false;
-  } else {
-    return true;
-  }
-}).t;
-
-const addListeners = (registeredFormatListeners: Cell<RegisteredFormats>, formats: string, callback: FormatChangeCallback, similar: boolean) => {
+const addListeners = (
+  editor: Editor,
+  registeredFormatListeners: Cell<RegisteredFormats>,
+  formats: string,
+  callback: FormatChangeCallback,
+  similar?: boolean,
+  vars?: FormatVars
+) => {
   const formatChangeItems = registeredFormatListeners.get();
 
   Arr.each(formats.split(','), (format) => {
-    if (!formatChangeItems[format]) {
-      formatChangeItems[format] = { similar, callbacks: [] };
-    }
+    const group = Obj.get(formatChangeItems, format).getOrThunk(() => {
+      const base: CallbackGroup = {
+        similarState: Cell(false),
+        similarCallbacks: [],
+        similarFalseState: Cell(false),
+        similarFalseCallbacks: [],
+        others: []
+      };
+
+      formatChangeItems[format] = base;
+      return base;
+    });
 
-    formatChangeItems[format].callbacks.push(callback);
+    const getCurrent = () => {
+      const parents = getParents(editor);
+      return matchingNode(editor, parents, format, similar, vars).isSome();
+    };
+    if (Type.isUndefined(vars)) {
+      if (similar) {
+        group.similarCallbacks.push(callback);
+        if (group.similarCallbacks.length === 1) {
+          group.similarState.set(getCurrent());
+        }
+      } else {
+        group.similarFalseCallbacks.push(callback);
+        if (group.similarFalseCallbacks.length === 1) {
+          group.similarFalseState.set(getCurrent());
+        }
+      }
```

Comment:
To go with Diogo's earlier comment, this logic is also duplicated so it would be good to refactor that.

Reply:
This one should be better now